haystack/finder.py

```diff
@@ -146,168 +147,194 @@ def eval(self, label_index: str = "feedback", doc_index: str = "eval_document",
-                        summed_f1_topk += best_f1
-                        summed_f1_topk_has_answer += best_f1
-
-                    if found_answer and found_em:
-                        break
-            # question not answerable
-            else:
-                number_of_no_answer += 1
-                # As question is not answerable, it is not clear how to compute average precision for this question.
-                # For now, we decided to calculate average precision based on the rank of 'no answer'.
-                for answer_idx, answer in enumerate(predicted_answers["answers"]):
-                    # check if 'no answer'
-                    if answer["answer"] is None:
-                        if answer_idx == 0:
-                            correct_no_answers_top1 += 1
-                            correct_readings_top1 += 1
-                            exact_matches_top1 += 1
-                            summed_f1_top1 += 1
-                        correct_no_answers_topk += 1
-                        correct_readings_topk += 1
-                        exact_matches_topk += 1
-                        summed_f1_topk += 1
-                        break
-        number_of_has_answer = correct_retrievals - number_of_no_answer
+            metric_counts = eval_counts_reader(question, predicted_answers, metric_counts)
 
         reader_total_time = time.time() - reader_start_time
+        number_of_has_answer = correct_retrievals - metric_counts["number_of_no_answer"]
+
         finder_total_time = time.time() - finder_start_time
 
         retriever_recall = correct_retrievals / number_of_questions
         retriever_map = summed_avg_precision_retriever / number_of_questions
 
-        reader_top1_accuracy = correct_readings_top1 / correct_retrievals
-        reader_top1_accuracy_has_answer = correct_readings_top1_has_answer / number_of_has_answer
-        reader_top_k_accuracy = correct_readings_topk / correct_retrievals
-        reader_topk_accuracy_has_answer = correct_readings_topk_has_answer / number_of_has_answer
-        reader_top1_em = exact_matches_top1 / correct_retrievals
-        reader_top1_em_has_answer = exact_matches_top1_has_answer / number_of_has_answer
-        reader_topk_em = exact_matches_topk / correct_retrievals
-        reader_topk_em_has_answer = exact_matches_topk_has_answer / number_of_has_answer
-        reader_top1_f1 = summed_f1_top1 / correct_retrievals
-        reader_top1_f1_has_answer = summed_f1_top1_has_answer / number_of_has_answer
-        reader_topk_f1 = summed_f1_topk / correct_retrievals
-        reader_topk_f1_has_answer = summed_f1_topk_has_answer / number_of_has_answer
-        reader_top1_no_answer_accuracy = correct_no_answers_top1 / number_of_no_answer
-        reader_topk_no_answer_accuracy = correct_no_answers_topk / number_of_no_answer
+        reader_top1_accuracy = metric_counts["correct_readings_top1"] / correct_retrievals
```

Comment:
Wouldnt it be possible to make a function out of this metric computations and putting it all inside the results dict?

And if we can do that, lets also use it for the eval_batch() method below.